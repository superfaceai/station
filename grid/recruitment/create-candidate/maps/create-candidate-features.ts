/* eslint-disable jest/no-export */
import { RecordingProcessOptions, SuperfaceTest } from '@superfaceai/testing';

import { buildSuperfaceTest } from '../../../test-config';

export const createCandidateFeaturesTest = (
  provider: string,
  options?: RecordingProcessOptions
): void => {
  describe(`recruitment/create-candidate/${provider}`, () => {
    describe('CreateCandidateFeatures', () => {
      let superface: SuperfaceTest;

      beforeAll(() => {
        superface = buildSuperfaceTest({
          profile: 'recruitment/create-candidate',
          provider,
          useCase: 'CreateCandidateFeatures',
        });
      });

      it(`returns valid ${provider} features`, async () => {
        const result = await superface.run(
          {
            input: {}
          },
          options
        );

        expect(result.isOk()).toBeTruthy();
        expect(result).toMatchSnapshot();
      });
    });
  });
};
