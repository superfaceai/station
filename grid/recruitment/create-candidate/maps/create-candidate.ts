/* eslint-disable jest/no-export, jest/valid-describe, jest/valid-title, jest/no-identical-title */

import { BinaryData } from '@superfaceai/one-sdk';
import { RecordingProcessOptions, SuperfaceTest } from '@superfaceai/testing';

import { buildSuperfaceTest } from '../../../test-config';

// eslint-disable-next-line @typescript-eslint/ban-types
type Candidate = {
  name: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  address: string;
  cv?: {
    fileName: string;
    data: BinaryData;
  };
  education: {
    degree: string;
    school: string;
    fieldOfStudy: string | null;
    startedAt: string;
    endedAt: string;
  }[];
  workExperience: {
    position: string;
    summary: string;
    company: string;
    industry: string;
    current: boolean;
    startedAt: string;
    endedAt: string;
  }[];
  links?: {
    name: string;
    url: string;
  }[];
};

const buildSampleCandidate = (): Candidate => {
  return {
    name: 'John Doe',
    firstName: 'John',
    lastName: 'Doe',

    email: 'john.doe@fakemail.com',
    phone: '1-859-557-6573',
    address: '25772 Gustave Shore, Iowa, USA',

    education: [
      {
        degree: 'MBA',
        school: 'University of Pennsylvania',
        fieldOfStudy: null,
        startedAt: '2008-03-01',
        endedAt: '2011-03-30',
      },
      {
        degree: 'B.S.',
        school: 'University of Chicago',
        fieldOfStudy: 'Marketing Communication & Economics',
        startedAt: '2004-09-01',
        endedAt: '2007-03-30',
      },
    ],

    workExperience: [
      {
        position: 'Sales Director',
        summary: 'Summary of a work experience at Test Company',
        company: 'Test Company',
        industry: 'Telecommunications',
        current: false,
        startedAt: '2011-03-01',
        endedAt: '2014-03-30',
      },
    ],

    cv: {
      fileName: 'cv-sample.pdf',
      data: BinaryData.fromPath('./grid/recruitment/cv-sample.pdf'),
    },

    links: [
      {
        name: 'twitter',
        url: 'https://twitter.com/candidate.username',
      },
      {
        name: 'Portfolio',
        url: 'https://url.to.portfolio',
      },
    ],
  };
};

const describeIf = (
  condition: boolean,
  ...args: Parameters<jest.Describe>
): void => (condition ? describe(...args) : describe.skip(...args));

export const createCandidateTest = (
  provider: string,
  jobIds: { valid: string; invalid: string },
  options?: RecordingProcessOptions
): void => {
  describe(`recruitment/create-candidate/${provider}`, () => {
    describe('CreateCandidate', () => {
      let superface: SuperfaceTest;
      let sampleCandidate: Candidate;

      beforeEach(() => {
        jest.setTimeout(10000);
        sampleCandidate = buildSampleCandidate();
        superface = buildSuperfaceTest({
          profile: 'recruitment/create-candidate',
          provider,
          useCase: 'CreateCandidate',
        });
      });

      describe('when specified job does exist', () => {
        describe('when unsupported CV file MIME type used', () => {
          it('returns CVMIMETypeNotSupported error', async () => {
            await expect(
              superface.run(
                {
                  input: {
                    jobId: jobIds.valid,
                    ...sampleCandidate,
                    cv: {
                      ...sampleCandidate.cv,
                      fileName: 'cv-sample.xml',
                    },
                  },
                },
                options
              )
            ).resolves.toMatchSnapshot();
          });
        });

        describe('when CV file name is missing', () => {
          it('returns CVFileNameRequired error', async () => {
            await expect(
              superface.run(
                {
                  input: {
                    jobId: jobIds.valid,
                    ...sampleCandidate,
                    cv: {
                      ...sampleCandidate.cv,
                      fileName: undefined,
                    },
                  },
                },
                options
              )
            ).resolves.toMatchSnapshot();
          });
        });

        describe('when valid inputs passed', () => {
          it('performs correctly', async () => {
            const result = await superface.run(
              {
                input: {
                  jobId: jobIds.valid,
                  ...sampleCandidate,
                },
              },
              options
            );
            expect(result.isOk()).toBeTruthy();
            expect(result).toMatchSnapshot();
          });
        });
      });

      describe('when specified job does not exist', () => {
        it('returns error', async () => {
          const result = await superface.run(
            {
              input: {
                jobId: jobIds.invalid,
                firstName: 'Demo',
                lastName: 'Testing',
                email: 'demo_testing@fakemail.com',
              },
            },
            options
          );

          expect(() => result.unwrap()).toThrow();
          expect(result).toMatchSnapshot();
        });
      });

      describeIf(
        provider === 'breezy-hr',
        'when specified company does not exist',
        () => {
          let companyId: string | undefined;

          beforeAll(() => {
            companyId = process.env.BREEZY_HR_COMPANY_ID;

            process.env.BREEZY_HR_COMPANY_ID = '1b111c1111ef11';
          });

          afterAll(() => {
            process.env.BREEZY_HR_COMPANY_ID = companyId;
          });

          it('returns error', async () => {
            const result = await superface.run(
              {
                input: {
                  jobId: jobIds.valid,
                  firstName: 'Demo',
                  lastName: 'Testing',
                  email: 'demo_testing@fakemail.com',
                },
              },
              options
            );

            expect(() => result.unwrap()).toThrow();
            expect(result).toMatchSnapshot();
          });
        }
      );

      describeIf(
        provider === 'workable',
        'when specified subdomain does not exist',
        () => {
          let subdomain: string | undefined;

          beforeAll(() => {
            subdomain = process.env.WORKABLE_SUBDOMAIN;

            process.env.WORKABLE_SUBDOMAIN = 'invalid-superface';
          });

          afterAll(() => {
            process.env.WORKABLE_SUBDOMAIN = subdomain;
          });

          it('returns error', async () => {
            const result = await superface.run(
              {
                input: {
                  jobId: jobIds.valid,
                  firstName: 'Demo',
                  lastName: 'Testing',
                  email: 'demo_testing@fakemail.com',
                },
              },
              options
            );

            expect(() => result.unwrap()).toThrow();
            expect(result).toMatchSnapshot();
          });
        }
      );
    });
  });
};
