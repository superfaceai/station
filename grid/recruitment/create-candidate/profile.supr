"""
Create Candidate
Create new candidate in ATS (Applicant Tracking System).
"""
name = "recruitment/create-candidate"
version = "2.0.0"

"""
Create Candidate
Creates an application for specific job. The candidate may (depends on the provider implementation) receive notification that her/his application was successfully accepted by the company.
"""
usecase CreateCandidate {

  input {
    """
    Job Id
    Id of a job, to which candidate applies to.
    """
    jobId! string!

    """
    First Name
    First name of a candidate.
    """
    firstName! string!

    """
    Last Name
    Last name of a candidate.
    """
    lastName! string!

    """
    Email
    Email of a candidate.
    """
    email! string!

    """
    Phone
    Phone number of a candidate.
    """
    phone string

    """
    Address
    Address of a candidate.
    """
    address string	

    """
    Education
    Array of education experiences.
    """
    education [Education]

    """
    Work Experience
    Array of work experiences.
    """
    workExperience [WorkExperience]

    """
    CV (Resume)
    Object containing either URL to a CV or data with CV document (BinaryData type) and fileName with the file name.
    """
    cv { url string } | { fileName string, data }

    """
    Links
    Array of links to different profiles or portfolios of a candidate.
    """
    links [Link]

    """
    Answers
    Array of answers to questions connected to a job that candidate applies to.
    """
    answers [Answer]

    """
    Stage Id
    Unique identifier of a stage, in which candidate is in.
    """
    stageId string
  }

  result {
    """
    Id
    Unique identifier of newly created candidate.
    """
    id! string!

    """
    Job Id
    Unique identifier of a job, newly created candidate applied for.
    """
    jobId! string!
    """
    Rate limit
    Rate limit details.
    """
    rateLimit RateLimit
  }!

  error Error

  example Successful {
    input {
      jobId = 'JOB_ID'
      firstName = 'Demo'
      lastName = 'Testing'
      email = 'demo_testing@fakemail.com'
    }

    result {
      id = 'CANDIDATE_ID'
      jobId = 'JOB_ID'
    }
  }

  example Failed {
    input {
      jobId = 'NOT_EXISTING_JOB_ID'
      firstName = 'Demo'
      lastName = 'Testing'
      email = 'demo_testing@fakemail.com'
    }

    error {
      title = 'Not found'
      detail = 'Not found'
      code = 'NotFound'
      rateLimit = {
        bucket = 'workable'
      }
    }
  }
}

"""
Get Create Candidate Features
Get features of CreateCandidate use case supported by the provider.
"""
usecase CreateCandidateFeatures {
  result {
    """
    CV MIME Types
    List of CV MIME types accepted by the provider. MIME types are defined in RFC2045 and registered by https://www.iana.org/.
    """
    cvMIMETypes! [string!]!
    """
    CV Upload Methods
    List of supported CV upload methods.
    """
    cvUploadMethods! [
      enum {
        """
        URL
        The cv input contains url field with publicly accessible URL of a CV file.
        """
        url = 'url'
        """
        File
        The cv input contains data field with CV file and fileName field with the file name.
        """
        file = 'file'
      }!
    ]!
 }

  example Successful {
    result {
      cvMIMETypes = ['application/pdf', 'text/rtf']
      cvUploadMethods = ['url', 'file']
    }
  }
}

"""
Education
Object representing educational experience of a candidate.
"""
model Education {
  """
  Id
  Unique identifier of candidate's education experience.
  """
  id string
  
  """
  School
  Name of a school.
  """
  school string

  """
  Degree
  Degree aquired by finishing candidate's education.
  """
  degree string

  """
  Field of Study
  Name of a field studied in candidate's education.
  """
  fieldOfStudy string
  
  """
  Started At
  ISO8601 timestamp when the education started.
  """
  startedAt string

  """
  Ended At
  ISO8601 timestamp when the education ended.
  """
  endedAt string
}

"""
Work Experience
Object representing work experience of a candidate.
"""
model WorkExperience {
  """
  Id
  Unique identifier of candidate's work experience.
  """
  id string

  """
  Position
  Name of position, candidate has occupied.
  """
  position string

  """
  Summary
  Summary of candidate's work experience. 
  """
  summary string

  """
  Company
  Name of company.
  """
  company string

  """
  Industry
  Name of industry, the company worked in.
  """
  industry string

  """
  Current
  Boolean representing whether candidate still work there.
  """
  current boolean

  """
  Started At
  ISO8601 timestamp when candidate started working.
  """
  startedAt string

  """
  Ended At
  ISO8601 timestamp when candidate stoped working.
  """
  endedAt string
}

"""
Answer
Object representing a answer to specific question from a job.
"""
model Answer {
  """
  Question Id
  Unique identifier of a question for this answer.
  """
  questionId string

  """
  Type
  Enum representing type of possible answer.
  """
  type enum {
    text
    textarea,
    number
    boolean
    multiple_choice
    multiple_select
    date
    file
  }

  """
  Value
  Value of an answer, is dependent on type of an answer.
  """
  value
}

"""
Link
Object representing a link specified by candidate.
"""
model Link {
  """
  Name
  Descriptive name of a link to some URL.
  """
  name string

  """
  URL
  Actual URL destination of the link.
  """
  url string
}

model Candidate {
  """
  Id
  Unique identifier of a candidate
  """
  id! string!
  
  """
  Job Id
  Unique identifier of a job, candidate applied for.
  """
  jobId! string!

  """
  Job Title
  Title of a job, candidate applied for.
  """
  jobTitle string

  """
  First Name
  First name of a candidate.
  """
  firstName! string!

  """
  Last Name
  Last name of a candidate.
  """
  lastName! string!

  """
  Full Name
  Full name of a candidate.
  """
  fullName string

  """
  Email
  Email of a candidate.
  """
  email! string!

  """
  Phone
  Phone number of a candidate.
  """
  phone string

  """
  Address
  Address of a candidate.
  """
  address string	

  """
  Links
  Array of links specified by a candidate. Can contain links to social profiles, portfolios or other.
  """
  links [Link]

  """
  Experiences
  Array of work experiences specified by a candidate.
  """
  experiences [WorkExperience]
  
  """
  Education
  Array of education experiences specified by a candidate.
  """
  education [Education]

  """
  CV (Resume)
  Object containing either url to a CV or data of a CV.
  """
  cv { url string } | { name string, data string }

  """
  CV Url
  Url pointing to CV of a candidate.
  """
  cvUrl string

  """
  Answers
  Array of answers to questions connected to a job that candidate applies to.
  """
  answers [Answer]

  """
  Stage
  Name of a stage, candidate is currently in.
  """
  stage string

  """
  Stage Changes
  Array of stages, candidate once was in (except for current one).
  """
  stageChanges [string]

  """
  Created At
  ISO8601 timestamp when was a candidate created.
  """
  createdAt string
  
  """
  Updated At
  ISO8601 timestamp when was a candidate lastly updated.
  """
  updatedAt string
}

model Error {
  """
  Error Title
  A short, human-readable summary of the problem type.
  """
  title!
  """
  Error Detail
  A human-readable explanation specific to this occurrence of the problem.
  """
  detail
  """
  Code
  Error code.
  """
  code! ErrorCode!
  """
  Rate limit
  Rate limit details.
  """
  rateLimit RateLimit
}

model RateLimit {
  """
  Bucket
  Different parts of API may have different rate limits.
  Bucket identifies to which part of API the rate limits apply.
  """
  bucket string
  """
  Total requests
  Total requests available in the time window.
  """
  totalRequests number
  """
  Remaining requests
  Remaining requests available in the time window.
  """
  remainingRequests number
  """
  Remaining requests available in percents
  Remaining requests available in the time window in percents.
  """
  remainingRequestsPercentage number
  """
  Reset timestamp
  Timestamp when the rate limits will reset (in  Unix Timestamp format). Available only for providers without rolling rate limit window.
  """
  resetTimestamp number
}

model ErrorCode enum {
  """
  Bad request
  The request was invalid or cannot be otherwise served.
  """
  BadRequest
  """
  Not found
  The profileId does not exist.
  """
  NotFound
  """
  Rate limit reached
  Returned when a request cannot be served due to the user or app reached max requests quota.
  """
  RateLimitReached
  """
  Unauthenticated
  There was a problem authenticating your request. Check that you are passing valid credentials.
  """
  Unauthenticated
  """
  Unauthorized
  The request is understood, but it has been refused or access is not allowed.
  """
  Unauthorized
  """
  Unknown error
  Returned when an unexpected error occured.
  """
  UnknownError
  """
  Conflict
  The request can not be executed, it is in conflict with current state of the target resource.
  """
  Conflict
  """
  CV file name required
  Missing cv.fileName input. CV file name is required when cv.data input field is defined.
  """
  CVFileNameRequired
  """
  CV MIME type not supported
  CV MIME type is not supported by the provider. To get supported MIME types run CreateCandidateFeatures use case.
  """
  CVMIMETypeNotSupported
}
