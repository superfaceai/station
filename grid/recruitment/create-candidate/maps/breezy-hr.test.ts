import nock from 'nock';

import { createCandidateTest } from './create-candidate';
import { createCandidateFeaturesTest } from './create-candidate-features';

createCandidateTest('breezy-hr', {
  valid: '701c3f2fc42a01',
  invalid: '12345ab789d1dd',
}, {
  beforeRecordingLoad: (recordings) => {
    recordings.forEach(recording => {
      if(recording.method === 'POST' && (/^\/v3\/company\/.+\/position\/.+\/candidate\/.+\/resume$/m.test(recording.path))) {
        const boundaryReplacement = '---boundary replaced';
        //resume upload contains multipart/form-data body with boundary with randomized value, we replace the boundary with constant in recorded body and matched body
        const recordedBody = replaceBoundaryInMultipartFormDataBody(boundaryReplacement, recording.body);
        recording.body = ( body => {
          const requestBody = replaceBoundaryInMultipartFormDataBody(boundaryReplacement, body);

          return recordedBody === requestBody;
        })
      }
    })
  }
});

createCandidateFeaturesTest('breezy-hr');

const replaceBoundaryInMultipartFormDataBody = ( (newBoundary: string, body?: nock.RequestBodyMatcher): string | undefined => {
  if (typeof(body) !== 'string') {
    return undefined;
  }
  const bodyString =  Buffer.from(body,'hex').toString();
  
  return bodyString.replace(/^(-*[0-9]+)/gm, newBoundary);
})
