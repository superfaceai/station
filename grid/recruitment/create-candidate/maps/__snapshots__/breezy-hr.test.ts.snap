// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recruitment/create-candidate/breezy-hr CreateCandidate when specified company does not exist returns error 1`] = `
Err {
  "error": "MappedError: Expected error
Properties: {
  \\"title\\": \\"Not Found\\",
  \\"detail\\": \\"Unable to get company by ID 1b111c1111ef11. Check that the company with that ID exists.\\",
  \\"code\\": \\"NotFound\\",
  \\"rateLimit\\": {
    \\"bucket\\": \\"breezy-hr\\",
    \\"totalRequests\\": 100,
    \\"remainingRequests\\": 92,
    \\"remainingRequestsPercentage\\": 92,
    \\"resetTimestamp\\": 1672411128
  }
}
AST Path: definitions[0].statements[1].statements[0]
Original Map Location: Line 13, column 5",
}
`;

exports[`recruitment/create-candidate/breezy-hr CreateCandidate when specified job does exist when CV file name is missing returns CVFileNameRequired error 1`] = `
Err {
  "error": "MappedError: Expected error
Properties: {
  \\"title\\": \\"Missing CV file name\\",
  \\"detail\\": \\"CV file name is required.\\",
  \\"code\\": \\"CVFileNameRequired\\"
}
AST Path: definitions[0].statements[0].statements[0]
Original Map Location: Line 6, column 5",
}
`;

exports[`recruitment/create-candidate/breezy-hr CreateCandidate when specified job does exist when unsupported CV file MIME type used returns CVMIMETypeNotSupported error 1`] = `
Err {
  "error": "MappedError: Expected error
Properties: {
  \\"title\\": \\"CV MIME type not supported\\",
  \\"detail\\": \\"File type cv-sample.xml is not supported by Breezy HR provider.\\",
  \\"code\\": \\"CVMIMETypeNotSupported\\"
}
AST Path: definitions[0].statements[0].statements[0]
Original Map Location: Line 6, column 5",
}
`;

exports[`recruitment/create-candidate/breezy-hr CreateCandidate when specified job does exist when valid inputs passed performs correctly 1`] = `
Ok {
  "value": Object {
    "id": "32004f13d65c01",
    "jobId": "701c3f2fc42a01",
    "rateLimit": Object {
      "bucket": "breezy-hr",
      "remainingRequests": 96,
      "remainingRequestsPercentage": 96,
      "resetTimestamp": 1672411128,
      "totalRequests": 100,
    },
  },
}
`;

exports[`recruitment/create-candidate/breezy-hr CreateCandidate when specified job does not exist returns error 1`] = `
Err {
  "error": "MappedError: Expected error
Properties: {
  \\"title\\": \\"Not Found\\",
  \\"detail\\": \\"Unable to get job by ID 12345ab789d1dd. Check that the job with that ID exists.\\",
  \\"code\\": \\"NotFound\\",
  \\"rateLimit\\": {
    \\"bucket\\": \\"breezy-hr\\",
    \\"totalRequests\\": 100,
    \\"remainingRequests\\": 93,
    \\"remainingRequestsPercentage\\": 93,
    \\"resetTimestamp\\": 1672411128
  }
}
AST Path: definitions[0].statements[2].statements[0]
Original Map Location: Line 21, column 5",
}
`;

exports[`recruitment/create-candidate/breezy-hr CreateCandidateFeatures returns valid breezy-hr features 1`] = `
Ok {
  "value": Object {
    "cvMIMETypes": Array [
      "application/pdf",
    ],
    "cvUploadMethods": Array [
      "file",
    ],
  },
}
`;
